### PR TITLE
[CPU EP] Add blocked quantization to QuantizeLinear op kernel

### PR DESCRIPTION
**Pull request by fajin-corp**
### Description
Add blocked quantization to QuantizeLinear op kernel.

If the quantize axis is not the last axis, block the tensor using 1x128 blocks. Blocks are dispatched to multiple threads for concurrently processing. Currently only support scalar instructions.

If the quantize axis is the last axis, block the tensor using 1 x quant_block_size blocks. Blocks are dispatched to multiple threads for concurrent processing. If output type is int types, call mlas kernel to use the SIMD instructions in each block.

#### Benchmark data
20 core 2GHz CPU, RelWithDebInfo config, 196 x 4096 tensor, quantize float to int4x2

Quantize before last axis:
 * single thread, scalar instruction: 31380900 ns
 * 8 thread, scalar instruction: 5098620 ns

Quantize last axis:
 * single thread, scalar instruction: 27927900 ns
 * 8 thread, SIMD instruction: 102261 ns

more thread, SIMD instruction, larger block size helps

### Motivation and Context
ONNX added blocked quantization to QuantizeLinear in optset 21

### CODE REVIEW

File: onnxruntime/core/util/qmath.h [hunk 1]
```diff
@@ -305,4 +305,591 @@
...
+                         const TOut* zero_point, TOut* output, std::ptrdiff_t M, std::ptrdiff_t K,
+                         const std::ptrdiff_t quant_block_size, bool saturate);
+};
+
+template <typename TOut>
+struct BlockedQuantizeLinear<float, TOut, 0> {
+  static void opNotLastAxis(concurrency::ThreadPool* thread_pool, const float* input, const float* scale,
+                            const TOut* zero_point, TOut* output, std::ptrdiff_t M, std::ptrdiff_t K,
+                            std::ptrdiff_t N, const std::ptrdiff_t quant_block_size,
+                            const std::ptrdiff_t block_size, bool saturate) {
```
Comment (yufenglee) [inline onnxruntime/core/util/qmath.h:369]: >block_size [](http://example.com/codeflow?start=49&length=10)

the naming of block_size and quant_block_size is a little bit confusing to me.

File: onnxruntime/core/util/qmath.h [hunk 2]
```diff
@@ -305,4 +305,591 @@
...
+    ORT_UNUSED_PARAMETER(saturate);
+    constexpr auto low = static_cast<int32_t>(std::numeric_limits<TOut>::lowest());
+    constexpr auto high = static_cast<int32_t>(std::numeric_limits<TOut>::max());
+    const auto num_r_block = (N + block_size - 1) / block_size;
+    const auto num_blocks = M * K * num_r_block;
```
Comment (yufenglee) [inline onnxruntime/core/util/qmath.h:374]: >M * K [](http://example.com/codeflow?start=28&length=5)

in real scenario, M * K is large usually. Maybe you don't need a block_size.

File: onnxruntime/core/util/qmath.h [hunk 3]
```diff
@@ -305,4 +305,591 @@
...
+    const TensorOpCost unit_cost{static_cast<double>(block_size * sizeof(float) * 2),
+                                 static_cast<double>(block_size * sizeof(TOut)),
+                                 static_cast<double>(block_size) * 2.0};
+    auto KN = K * N;
+    auto qKN = (K + quant_block_size - 1) / quant_block_size * N;
+    const auto Kbc = K * num_r_block;
```
Comment (yufenglee) [inline onnxruntime/core/util/qmath.h:380]: nit: could we use a better name

File: onnxruntime/core/util/qmath.h [hunk 4]
```diff
@@ -305,4 +305,591 @@
...
+template <typename TOut>
+struct BlockedQuantizeLinear<float, TOut, 0> {
+  static void opNotLastAxis(concurrency::ThreadPool* thread_pool, const float* input, const float* scale,
+                            const TOut* zero_point, TOut* output, std::ptrdiff_t M, std::ptrdiff_t K,
+                            std::ptrdiff_t N, const std::ptrdiff_t quant_block_size,
+                            const std::ptrdiff_t block_size, bool saturate) {
+    ORT_UNUSED_PARAMETER(saturate);
+    constexpr auto low = static_cast<int32_t>(std::numeric_limits<TOut>::lowest());
+    constexpr auto high = static_cast<int32_t>(std::numeric_limits<TOut>::max());
+    const auto num_r_block = (N + block_size - 1) / block_size;
```
Comment (yufenglee) [inline onnxruntime/core/util/qmath.h:373]: >r [](http://example.com/codeflow?start=19&length=1)

nit: what does r represent here?

File: onnxruntime/core/util/qmath.h [hunk 5]
```diff
@@ -305,4 +305,591 @@
...
+    const auto Kbc = K * num_r_block;
+
+    concurrency::ThreadPool::TryParallelFor(
+        thread_pool,
+        num_blocks,
+        unit_cost,
+        [&](std::ptrdiff_t begin, std::ptrdiff_t end) {
+          auto m = begin / Kbc, k = begin % Kbc / num_r_block, nb = begin % num_r_block, n = nb * block_size;
+          auto output_idx = m * KN + k * N + n;
+          auto zp_b_idx = m * qKN + k / quant_block_size * N;
```
Comment (yufenglee) [inline onnxruntime/core/util/qmath.h:389]: >zp_b_idx [](http://example.com/codeflow?start=15&length=8)

nit: quant_param_idx

File: onnxruntime/core/util/qmath.h (hunk 1 above)
Comment (fajin-corp) [reply onnxruntime/core/util/qmath.h:369]: renamed to `thread_block_size`

File: onnxruntime/core/util/qmath.h (hunk 4 above)
Comment (fajin-corp) [reply onnxruntime/core/util/qmath.h:373]: changed to `row`

File: onnxruntime/core/util/qmath.h (hunk 2 above)
Comment (fajin-corp) [reply onnxruntime/core/util/qmath.h:374]: this is to calculate number of thread blocks, so the number depends on the thread block size.

File: onnxruntime/core/util/qmath.h (hunk 4 above)
Comment (fajin-corp) [reply onnxruntime/core/util/qmath.h:373]: I guess there is a confusion between quant block and thread block. Let me differentiate these two concepts using naming.

File: onnxruntime/core/util/qmath.h (hunk 3 above)
Comment (fajin-corp) [reply onnxruntime/core/util/qmath.h]: updated the naming to differentiate thread block and quantize block

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**